lib/dry/result_matcher/evaluator.rb

```diff
@@ -0,0 +1,40 @@
+module Dry
+  module ResultMatcher
+    class Evaluator
+      attr_reader :result
+      attr_reader :output
+
+      def initialize(result, cases)
+        @result = result
+        @output = nil
+        @matched = false
+
+        cases.each do |name, kase|
+          define_singleton_method(name) do |pattern = nil, &block|
+            handle_case(kase, pattern, &block)
+          end
+        end
```

Comment:
This will *probably* have a pretty serious impact on performance, it would be better to define methods on classes, it's the same issue as with using `extend` on instances at runtime.